keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,1699 @@
+currently no meaningful way to support these technologies in Kubernetes. This
+proposal suggests to address the issue above in a generalized way by extending
+the Kubernetes resource model with a new type of resources, i.e. QoS-class
+resources.
+
+This KEP identifies two technologies that can immediately be enabled with
+QoS-class resources. However, these are just two examples  and the proposed
+changes are generic (and not tied to these two QoS-class resource types in any
+way), making it easier to implement new QoS-class resource types.
+
+[Intel RDT][intel-rdt] implements a class-based mechanism for controlling the
+cache and memory bandwidth QoS of applications. All processes in the same
+hardware class share a portion of cache lines and memory bandwidth. RDT
+provides a way for mitigating noisy neighbors and fulfilling SLAs. In Linux
+control happens via resctrl -- a pseudo-filesystem provided by the kernel which
+makes it virtually agnostic of the hardware architecture. The OCI runtime-spec
+has supported Intel RDT for a while already. Other hardware vendors have
+comparable technologies which use the same [resctrl interface][linux-resctrl].
+
+The Linux Block IO controller parameters depend very heavily on the underlying
+hardware and system configuration (device naming/numbering, IO scheduler
+configuration etc) which makes it very impractical to control from the Pod spec
+level. In order to hide this complexity the concept of blockio classes has been
+added to the container runtimes (CRI-O and containerd). A system administrator
+is able to configure blockio controller parameters on per-class basis and the
+classes are then made available for CRI clients. Following this model also
+provides a possible framework for the future improvements, for instance
+enabling class-based network or memory type prioritization of applications.
+
+Currently, there is no mechanism in Kubernetes to use these types of resources.
+CRI-O and containerd runtimes have support for RDT and blockio classes and they
+provide an bridge-gap user interface through special pod annotations. The goal
+is to get these types of resources first class citizens and properly supported
+in Kubernetes, providing visibility, a well-defined user interface, and
+permission controls.
+
+It seems necessary to support both container-level and pod-level QoS-class
+resources as independent concepts. Intel RDT (above) is per-container by
+design because of the hardware implementation (the control/class hiearchy is
+flat). Also, the current support for blockio is container-level only (it is not
+possible to configure pod sandbox-level cgroup parameters). However, having
+pod-level QoS-class resources makes it possible to implement support for
+sandbox-level blockio parameters. Other usage for pod sandbox-level QoS-class
+resources would be communicating the Kubernetes Pod QoS class from
+kubelet to the container runtime.
+
+### Goals
```

Comment:
I tried to put all considerations for new resources into this list of questions. I can find some answers in this KEP, but not other. Can you please think of the way (maybe separate section) to answer these questions:

1. Advertising resources
  - Are QoS resources static for the lifetime of a Node?
  - How kubelet will know about resource availability?
  - If not static, how will kubelet be notified of a change to start eviction or start admitting pods?
  - Will there be any “system reserved” portion of QoS resources that kubelet needs to be aware of?
  - Would it be often that static Pods needs QoS resources? (note: static pod scheduling sometimes makes race with other resources scheduling)

2. Scheduling Pods with QoS resources
  - How scheduler will make a decision on node assignment based on QoS resource availability? Is it fine to just take first that will fit or scheduler typically will need to spread out Pods with the same QoS?
  - How different classes of QoS must be spread out? 

3. Limits, quotas, and defaults
  - Are those resources subject for limit ranger, defaults, and quotas?

4. Admission on the node
  - What information will the kubelet need to know to admit the Pod with the specific QoS? Will it require additional external calls to the runtime?
  - Will pod admission fall in a situation when Pod seems to be scheduleable on the Node based on QoS, but admission fails? How many additional factors pod admission may take into consideration when dealing with pod admission and allocating of the resource?

5. Allocating resource
  - If kubelet restart, will the QoS class assignment be in-tact or there needs to be an additional state recording Pod to resource allocation?

6. Support overprovisioning
  - How overprovisioning may be supported by this resource? Will there be a setting to allow “better” class unless it’s taken by a new Pod?
  - Will bursting in other QoS class be desired?
  - Will there be a need to dynamically change class assignment based on other Pods being scheduled on the Node?

7. Monitoring of usage
  - How do end users know that the resource is being used properly - correct QoS is selected or better one is needed?
  - Does kubelet need to have any monitoring of resource usage beyond # of pods using the resource?

8. Eviction
  - Will we need to think of an eviction of Pods based on classes availability?
  - Will limits of resources require kubelet evict lower priority Pods to allow higher priority pods to be scheduled?

Reply:
I don't know if it was evident but I replied to these in a separate comment (iirc there was some problem back then and I wasn't able to reply to this comment directly back then)
https://github.com/kubernetes/enhancements/pull/3004#issuecomment-1414043336